base absdiff() off of recent changes to abs()

blitmap opened the pull request:
Hi, the other guy (@jeff-ayars) got to abs() before me :-)  There was a comment about updating absdiff() to reflect the new definition of abs() - that is what I have done here.

(g'day!)

PS: I wasn't thinking when I made a pull request off of my master, now this includes a cleanup to p_median() also.

6thimage commented:
Your p_median contains an error that the original does - after line 25, it should use the sort variable rather than a (otherwise it is ignoring the sort).

Additionally, you could remove the temp variable and just use c throughout the function - I'm guessing the compiler would probably do this when optimising anyway.

blitmap commented:
@6thimage:

1) I believe I have corrected the original issue now, and changed the stack allocation of the sorted vector to be on the heap - as I am not sure how long the vector might be (my reasoning).  I modeled it after the allocation in mode() as an example for this.

2) I did declare `tmp` thinking it would give the compiler more freedom to do its thing.  It probably produces the same code though...

Do you approve?  I can still remove `tmp` if it goes against a style guideline or something ~

blitmap commented:
I am not sure the test likes a heap-allocated vector - I modelled this after p_mode() but that function doesn't have a test.  I"m either doing something wrong or...

6thimage commented:
b5df390 looks like it has sorted everything - a compiler will more than likely optimise out the tmp variable, to me it just looks a little cleaner without the tmp variable. As to giving the compiler more freedom, the only thing I can see is that a dumb compiler might use an additional register, impacting performance slightly.

The test is failing because p_malloc is currently not implemented - have a look at src/base/p_malloc.c.

blitmap commented:
Now I think the test is failing because it hates me.  Or the test was created against the old behavior... which was incorrect.  Not sure.
